src/utils/streams.js

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+// This file holds various utilities about streams.
+
+import { Transform, Writable } from 'stream';
+import crypto from 'crypto';
+
+import { getLogger, type Logger } from '../log';
+
+/**
+ * This transform computes a sha1 of the daa passing through it, but otherwise
+ * doesn't alter the data.
+ */
+export class HasherPassThrough extends Transform {
+  log: Logger = getLogger('HasherPassThrough');
+  sha1Hasher: crypto$Hash = crypto.createHash('sha1');
+  sha1Result: string | null = null;
+
+  _transform(chunk: *, encoding: *, callback: *) {
+    this.sha1Hasher.update(chunk);
+    this.push(chunk);
+    callback();
+  }
+
+  sha1() {
+    this.log.verbose('sha1()');
+    if (this.sha1Result === null) {
+      this.sha1Result = this.sha1Hasher.digest('hex');
+    }
+    return this.sha1Result;
+  }
+}
+
+/**
+ * This transform's sole goal is to check that the size of the streamed data
+ * stays within some limit.
+ */
+export class LengthCheckerPassThrough extends Transform {
+  log: Logger = getLogger('LengthCheckerPassThrough');
+  maxLength: number;
+  length: number = 0;
+
+  constructor(maxLength: number) {
+    super();
+    this.maxLength = maxLength;
+  }
+
+  _transform(chunk: *, encoding: *, callback: *) {
+    this.length += chunk.length;
+    this.log.verbose(
+      'length-checker-length',
+      `Added chunk's length is ${chunk.length}, current length is ${this.length}`
+    );
+
+    if (this.length > this.maxLength) {
+      // Using debug instead of verbose will make it possible to assert it in
+      // tests that we actually checked the length through this class.
+      this.log.debug('length-checker-length-error');
+      callback(
+        new Error(
+          `The length is bigger than the configured maximum ${this.maxLength}.`
+        )
+      );
+      return;
+    }
+
+    this.push(chunk);
+    callback();
+  }
+}
+
+/**
+ * This writable stream keeps all received chunks until the stream is closed.
+ * Then the chunks are concatenated into a unique Buffer that can be retrieved.
+ */
+export class Concatenator extends Writable {
```

Comment:
Ah, that's exactly what I intend to do by moving away from the hash-as-filename system :-) ([link](https://docs.google.com/document/d/1rWfq2N7ie_XzGELq217QIJrOF500_BwHJFoauPY2Ayg/edit)).

Another solution could be to 1. upload the file to a temp name, and 2. rename it to the computed hash. But Google counts this as 2 write operations (twice the price!) so I decided to implement the first solution instead. Also maybe this would be more prone to races. I don't know really.

Also that's why I limited the upload size to 32MB like they do App Engine for now. But I'm worried like you are. I don't know what the limitations are on the production environment.

Reply:
I filed #31